smart-queue/enqueue.js

```diff
@@ -0,0 +1,78 @@
+import * as QueueActionTypes from "./action-types";
+
+function validKey(key) {
+  if (key === null) {
+    console.log("Missing key, every queue action should have a key!");
+    return false;
+  }
+  return true;
+}
+
+function indexOfAction(array, key) {
+  return array.findIndex(
+    ({ meta: { offline: { queue = null } } }) => queue && queue.key === key
+  );
+}
+export function enqueue(array, action) {
+  const { meta: { offline: { queue = null } } } = action;
+  let index, existingAction;
+
+  if (!queue) {
+    return [...array, action];
+  }
+
+  const { method = "UNKNOWN", key = null } = queue;
+  if (!validKey(key)) {
+    return array;
+  }
+
+  index = indexOfAction(array, key);
+
+  switch (method) {
+    case QueueActionTypes.CREATE:
+      if (index !== -1) {
+        array[index] = { ...array[index], ...action };
+        return array;
+      }
+      return [...array, action];
+    case QueueActionTypes.DELETE:
+      if (index !== -1) {
+        existingAction = array[index];
+        if (
+          existingAction.meta.offline.queue.method ===
+            QueueActionTypes.UPDATE ||
+          existingAction.meta.offline.queue.method === QueueActionTypes.CREATE
+        ) {
+          array.splice(index, 1);
+        }
+      }
+      return array;
+    case QueueActionTypes.UPDATE:
+      if (index !== -1) {
+        existingAction = array[index];
+        if (
+          existingAction.meta.offline.queue.method === QueueActionTypes.CREATE
+        ) {
+          array[index] = { ...array[index], ...action };
```

Comment:
This merge is shallow, so you'd most likely be entirely replacing the create action with the update action. Running an update action when the create action has likely not been made yet seems problematic.

I imagine you would want to replace only the data/payload of `createAction.meta.offline.effect` and leave everything else as is. How you would perform this merge would be dependent on the user's effect reconciler.

Reply:
If we update `data` part (`createAction.meta.offline.effect.body`) only, then should we not consider `meta` portion of `commit` and `rollback` as well? It might happen that user wants to pass updated `data` to `commit` or `rollback` reducer.

Reply:
That is true.